### PR TITLE
M5608 Row-level security

### PR DESCRIPTION
**Pull request by dennishendriksen**
#### Checklist
- [x] Functionality works & meets specifications
- [x] Code reviewed
- [x] Code unit/integration/system tested
- (n.a.) User documentation updated (_postponed untill we have new UI_)
- (n.a.) (If you have changed REST API interface) view-swagger.ftl updated
- [x] Test plan template updated
- [x] Clean commits (_commits will be squashed_)
- [x] Added Feature/Fix to release notes
- [x] Integration tests run correctly

### CODE REVIEW

File: molgenis-bootstrap/src/main/java/org/molgenis/bootstrap/MolgenisBootstrapper.java [hunk 1]
```diff
@@ -98,6 +97,8 @@ public void onApplicationEvent(ContextRefreshedEvent event)
 		systemEntityTypeBootstrapper.bootstrap(event);
 		LOG.debug("Bootstrapped system entity meta data");
 
+		bootstrappingEventPublisher.publishBootstrappingSystemEntitiesFinishedEvent();
```
Comment (dennishendriksen) [inline molgenis-bootstrap/src/main/java/org/molgenis/bootstrap/MolgenisBootstrapper.java]: Is this change still required?

File: molgenis-bootstrap/src/main/java/org/molgenis/bootstrap/MolgenisBootstrapper.java [hunk 2]
```diff
@@ -129,7 +130,7 @@ public void onApplicationEvent(ContextRefreshedEvent event)
 		bootstrapThemePopulator.populate();
 		LOG.debug("Populated bootstrap themes");
 
-		bootstrappingEventPublisher.publishBootstrappingFinishedEvent();
+		bootstrappingEventPublisher.publishBootstrappingSystemEntitiesFinishedEvent();
```
Comment (dennishendriksen) [inline molgenis-bootstrap/src/main/java/org/molgenis/bootstrap/MolgenisBootstrapper.java]: Is this change still required?

File: molgenis-core-ui/src/main/java/org/molgenis/core/ui/admin/permission/PermissionManagerController.java [hunk 3]
```diff
@@ -221,6 +292,56 @@ public void updateUserEntityClassPermissions(@RequestParam String userId, WebReq
...
+	{
+		EntityType entityType = dataService.getEntityType(entityTypeRlsRequest.getId());
+		String aclClassType = toAclClassType(entityType);
+		boolean hasAclClass = mutableAclClassService.hasAclClass(aclClassType);
+		if (entityTypeRlsRequest.isRlsEnabled())
+		{
+			if (!hasAclClass)
+			{
+				mutableAclClassService.createAclClass(aclClassType, toIdClass(entityType));
+				// TODO create ACLs for existing entities
```
Comment (dennishendriksen) [inline molgenis-core-ui/src/main/java/org/molgenis/core/ui/admin/permission/PermissionManagerController.java]: Resolve TODO

File: molgenis-core-ui/src/main/java/org/molgenis/core/ui/admin/permission/PermissionManagerController.java [hunk 4]
```diff
@@ -221,6 +292,56 @@ public void updateUserEntityClassPermissions(@RequestParam String userId, WebReq
...
+			case LONG:
+				return Long.class;
+			default:
+				throw new UnexpectedEnumException(attributeType);
+		}
+	}
+
+	private static String toAclClassType(EntityType entityType)
+	{
+		return "entity-" + entityType.getId();
```
Comment (dennishendriksen) [inline molgenis-core-ui/src/main/java/org/molgenis/core/ui/admin/permission/PermissionManagerController.java]: Code duplication: move to utility class

File: molgenis-data-security/src/main/java/org/molgenis/data/security/acl/SidUtils.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import org.springframework.security.acls.domain.GrantedAuthoritySid;
+import org.springframework.security.acls.domain.PrincipalSid;
+import org.springframework.security.acls.model.Sid;
+
+/**
+ * Util class to create security identities for {@link User users} and {@link Group groups}.
+ *
+ * @see Sid
+ */
+public class SidUtils
```
Comment (dennishendriksen) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/acl/SidUtils.java]: Already exists in molgenis-security module. Duplicated?

File: molgenis-data-security/src/main/java/org/molgenis/data/security/owned/RowLevelSecurityRepositoryDecorator.java [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+import java.util.stream.Stream;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+import static org.molgenis.data.security.EntityIdentity.TYPE_PREFIX;
+import static org.molgenis.data.security.EntityTypePermission.READ;
+import static org.molgenis.data.security.EntityTypePermission.WRITE;
+
+/**
+ * RepositoryDecorator that works on EntityType that extends OwnedEntityType.
```
Comment (dennishendriksen) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/owned/RowLevelSecurityRepositoryDecorator.java]: Update documentation

File: molgenis-data-security/src/main/java/org/molgenis/data/security/owned/RowLevelSecurityRepositoryDecorator.java [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+	@Override
+	public Query<Entity> query()
+	{
+		return new QueryImpl<>(this);
+	}
+
+	@Override
+	public Iterator<Entity> iterator()
+	{
+		if (isRowLevelSecured())
```
Comment (dennishendriksen) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/owned/RowLevelSecurityRepositoryDecorator.java]: Add curly braces

File: molgenis-data-security/src/main/java/org/molgenis/data/security/owned/RowLevelSecurityRepositoryDecorator.java [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+	}
+
+	@Override
+	public void forEachBatched(Fetch fetch, Consumer<List<Entity>> consumer, int batchSize)
+	{
+		delegate().forEachBatched(fetch, entities ->
+		{
+			if (isRowLevelSecured())
+			{
+				//TODO: This results in smaller batches! Should do a findAll instead!
```
Comment (dennishendriksen) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/owned/RowLevelSecurityRepositoryDecorator.java]: Resolve TODO

File: molgenis-data-security/src/main/java/org/molgenis/data/security/owned/RowLevelSecurityRepositoryDecorator.java [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+				return true;
+			}));
+		}
+		return delegate().add(entities);
+	}
+
+	private boolean isRowLevelSecured()
+	{
+		// TODO 'entity-' code duplication
+		return mutableAclClassService.hasAclClass(TYPE_PREFIX + getName());
```
Comment (dennishendriksen) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/owned/RowLevelSecurityRepositoryDecorator.java]: Code duplication, move to utility class (see PermissionManagerController)

File: molgenis-data/src/main/java/org/molgenis/data/decorator/DynamicRepositoryDecoratorRegistryImpl.java [hunk 10]
```diff
@@ -100,6 +100,6 @@ public DynamicRepositoryDecoratorFactory getFactory(String id)
 	@EventListener
 	public void onApplicationEvent(BootstrappingEvent bootstrappingEvent)
 	{
-		this.bootstrappingDone = bootstrappingEvent.getStatus() == FINISHED;
+		this.bootstrappingDone = bootstrappingEvent.getStatus() == FINISHED_SYSTEM_ENTITY_TYPES;
```
Comment (dennishendriksen) [inline molgenis-data/src/main/java/org/molgenis/data/decorator/DynamicRepositoryDecoratorRegistryImpl.java]: Changes required?

File: molgenis-data/src/main/java/org/molgenis/data/event/BootstrappingEvent.java [hunk 11]
```diff
@@ -4,7 +4,7 @@
 {
 	public enum BootstrappingStatus
 	{
-		STARTED, FINISHED;
+		FINISHED_SYSTEM_ENTITY_TYPES;
```
Comment (dennishendriksen) [inline molgenis-data/src/main/java/org/molgenis/data/event/BootstrappingEvent.java]: Changes required?

File: molgenis-data/src/main/java/org/molgenis/data/event/BootstrappingEventPublisher.java [hunk 12]
```diff
@@ -18,13 +17,8 @@ public BootstrappingEventPublisher(ApplicationEventPublisher publisher)
 		this.publisher = publisher;
 	}
 
-	public void publishBootstrappingStartedEvent()
+	public void publishBootstrappingSystemEntitiesFinishedEvent()
```
Comment (dennishendriksen) [inline molgenis-data/src/main/java/org/molgenis/data/event/BootstrappingEventPublisher.java]: Changes required?

File: molgenis-data/src/test/java/org/molgenis/data/decorator/DynamicRepositoryDecoratorRegistryImplTest.java [hunk 13]
```diff
@@ -65,7 +65,7 @@ public void testDecorate()
 		DynamicRepositoryDecoratorRegistryImpl dynamicRepositoryDecoratorRegistry = new DynamicRepositoryDecoratorRegistryImpl(
 				dataService);
 		//fake the event to tell the registry that bootstrapping is done.
-		dynamicRepositoryDecoratorRegistry.onApplicationEvent(new BootstrappingEvent(FINISHED));
+		dynamicRepositoryDecoratorRegistry.onApplicationEvent(new BootstrappingEvent(FINISHED_SYSTEM_ENTITY_TYPES));
```
Comment (dennishendriksen) [inline molgenis-data/src/test/java/org/molgenis/data/decorator/DynamicRepositoryDecoratorRegistryImplTest.java]: Changes required?

File: molgenis-data/src/test/java/org/molgenis/data/decorator/DynamicRepositoryDecoratorRegistryImplTest.java [hunk 14]
```diff
@@ -82,7 +82,7 @@ public void testDecorateNoDecorator()
...
```
Comment (dennishendriksen) [inline molgenis-data/src/test/java/org/molgenis/data/decorator/DynamicRepositoryDecoratorRegistryImplTest.java]: Changes required?

File: molgenis-platform-integration-tests/src/test/java/org/molgenis/integrationtest/platform/PlatformITConfig.java [hunk 15]
```diff
@@ -238,7 +237,7 @@ public void onApplicationEvent(ContextRefreshedEvent event)
 					event.getApplicationContext().getBean(EntityTypeRegistryPopulator.class).populate();
 					event.getApplicationContext().getBean(DynamicDecoratorPopulator.class).populate();
 
-					bootstrappingEventPublisher.publishBootstrappingFinishedEvent();
+					bootstrappingEventPublisher.publishBootstrappingSystemEntitiesFinishedEvent();
```
Comment (dennishendriksen) [inline molgenis-platform-integration-tests/src/test/java/org/molgenis/integrationtest/platform/PlatformITConfig.java]: Changes required?

File: molgenis-security/src/main/java/org/molgenis/security/permission/UserPermissionEvaluatorImpl.java [hunk 16]
```diff
@@ -15,16 +17,21 @@
...
 		this.permissionEvaluator = requireNonNull(permissionEvaluator);
+		this.mutableAclService = requireNonNull(mutableAclService);
+		this.userAccountService = requireNonNull(userAccountService);
 	}
 
 	@Override
 	public boolean hasPermission(ObjectIdentity objectIdentity, Permission permission)
 	{
-		if (SecurityUtils.currentUserIsSuOrSystem())
+		if (SecurityUtils.currentUserIsSuOrSystem() || isOwner(objectIdentity))
```
Comment (dennishendriksen) [inline molgenis-security/src/main/java/org/molgenis/security/permission/UserPermissionEvaluatorImpl.java]: Change required? Isn't this handled by Spring Security ACL?

File: molgenis-security/src/test/java/org/molgenis/security/permission/UserPermissionEvaluatorImplTest.java [hunk 17]
```diff
@@ -32,7 +32,8 @@
 	@BeforeMethod
 	public void setUpBeforeMethod()
 	{
-		userPermissionEvaluator = new UserPermissionEvaluatorImpl(permissionEvaluator);
+		//TODO: fix + add "ownership tests"
```
Comment (dennishendriksen) [inline molgenis-security/src/test/java/org/molgenis/security/permission/UserPermissionEvaluatorImplTest.java]: Resolve TODO

**Comment by dennishendriksen** [comment]
Todo: Update PostgreSqlExceptionTranslator which currently assumes all tables are entity type tables (acl tables are not entity type tables)

File: molgenis-core-ui/src/main/java/org/molgenis/core/ui/admin/permission/PermissionManagerController.java [hunk 18]
```diff
@@ -221,6 +292,56 @@ public void updateUserEntityClassPermissions(@RequestParam String userId, WebReq
 		updateEntityTypePermissions(webRequest, sid);
 	}
 
+	@PreAuthorize("hasAnyRole('ROLE_SU')")
+	@Transactional
+	@PostMapping("/update/entityclass/rls")
+	@ResponseStatus(HttpStatus.OK)
+	public void updateEntityClassRls(@Valid @RequestBody EntityTypeRlsRequest entityTypeRlsRequest)
```
Comment (dennishendriksen) [inline molgenis-core-ui/src/main/java/org/molgenis/core/ui/admin/permission/PermissionManagerController.java]: Todo: add unit tests

**Comment by dennishendriksen** [comment]
TODO enable RLS for previously owned system entity types

**Comment by dennishendriksen** [comment]
TODO changing RLS for system entity types not possible at runtime (only during bootstrapping)

File: molgenis-data-security/src/main/java/org/molgenis/data/security/meta/SecurityMetadataPackage.java [hunk 19]
```diff
@@ -0,0 +1,33 @@
+package org.molgenis.data.security.meta;
```
Comment (dennishendriksen) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/meta/SecurityMetadataPackage.java]: TODO can be removed

File: molgenis-data-security/src/main/java/org/molgenis/data/security/EntityIdentity.java [hunk 20]
```diff
@@ -0,0 +1,18 @@
+package org.molgenis.data.security;
+
+import org.molgenis.data.Entity;
+import org.molgenis.data.meta.model.EntityType;
+import org.springframework.security.acls.domain.ObjectIdentityImpl;
+
+public class EntityIdentity extends ObjectIdentityImpl
+{
+	public EntityIdentity(EntityType entityType, Entity entity)
```
Comment (dennishendriksen) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/EntityIdentity.java]: Remove EntityType argument, entity type can be retrieved by calling entity.getEntityType

File: molgenis-data-platform/src/main/java/org/molgenis/data/meta/system/SystemEntityTypePersister.java [hunk 21]
```diff
@@ -56,6 +61,21 @@ public void persist()
 		// persist EntityType entities
 		List<EntityType> metaEntityMetaSet = systemEntityTypeRegistry.getSystemEntityTypes().collect(toList());
 		injectExistingEntityTypeAttributeIdentifiers(metaEntityMetaSet);
+		metaEntityMetaSet.forEach(systemEntityType ->
```
Comment (dennishendriksen) [inline molgenis-data-platform/src/main/java/org/molgenis/data/meta/system/SystemEntityTypePersister.java]: TODO add unit tests

File: molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/QuestionnaireMetaData.java [hunk 22]
```diff
@@ -48,6 +42,7 @@ public void init()
 		}
 
 		addAttribute(ATTR_STATUS).setDataType(ENUM).setEnumOptions(enumOptions).setVisible(false).setNillable(false);
+		addAttribute(OWNER_USERNAME).setDataType(STRING).setVisible(false).setNillable(false);
```
Comment (dennishendriksen) [inline molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/QuestionnaireMetaData.java]: TODO fix questionnaire plugin

File: molgenis-data-security/src/main/java/org/molgenis/data/security/owned/RowLevelSecurityRepositoryDecorator.java (hunk 6 above)
Comment (dennishendriksen) [reply molgenis-data-security/src/main/java/org/molgenis/data/security/owned/RowLevelSecurityRepositoryDecorator.java]: Write unit tests

File: molgenis-data-security/src/main/java/org/molgenis/data/security/owned/RowLevelSecurityRepositoryDecorator.java [hunk 23]
```diff
@@ -0,0 +1,293 @@
...
+		return delegate().add(entities);
+	}
+
+	private boolean isRowLevelSecured()
+	{
+		String aclClass = EntityIdentityUtils.toType(getEntityType());
+		return mutableAclClassService.hasAclClass(aclClass);
+	}
+
+	private boolean hasPermissionOnEntity(Entity entity, EntityTypePermission permission)
```
Comment (dennishendriksen) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/owned/RowLevelSecurityRepositoryDecorator.java]: TODO should user EntityPermission, nog EntityTypePermission

**Comment by dennishendriksen** [comment]
TODO handle the case were a user exists with username SYSTEM

**Comment by dennishendriksen** [comment]
Travis says ok, please ignore continuous-integration/travis-ci/push failure

**Comment by tommydeboer** [comment]
**Issue:**
Uploading a dataset as a _user_ but activating row level security as _admin_ will make the admin owner of the rows instead of the user.

**Comment by bartcharbon** [comment]
The form validation does not take rows that you cannot see into account:
- admin has a row with id "test"
- if a user creates a new row with id "test" it fails on save, form validation does not mark the field as illegal

**Comment by tommydeboer** [comment]
What should be the behaviour when referenced entity types are row level secured but the referencing entity type is not? 

Using this case:
1. Import it_emx_datatypes as admin
2. Give user WRITE permission to TypeTest
3. Give user READ permission to TypeTestRef, Location and Person
4. Activate RLS for TypeTestRef
5. Log in as user
6. Go to TypeTest

Reference values are not shown in the table, which is correct I think. But the user can't edit the row:

```entity [it_emx_datatypes_TypeTestRef] with value [ref2] does not exist```

---------

What if I now give the user permission to add rows to the referenced entity type?

1. Give user WRITE permission to TypeTestRef
2. Log in as user
3. Go to TypeTestRef
4. Add a single row  'testref'
5. Go to TypeTest
6. Click 'add row'

The categoricals, entity type select boxes etc. all only show the 'testref' entity, which is good.

7. Select the checkbox 'testref' for 'xcategoricalmref_value '

This triggers an error:
```Uncaught TypeError: Cannot read property 'label' of undefined```
But you can continue adding the row.

8. Finish entering only the required fields and click create
9. Try to edit the newly added row

This still gives the error:
```entity [it_emx_datatypes_TypeTestRef] with value [ref2] does not exist```
Even though 'ref2' is invisible to the user, it is still added as a **default value**

------

1. Add a new row, this time filling out all the fields
2. Try to edit the row

This still gives an error:
```entity [it_emx_datatypes_TypeTestRef] with value [ref1] does not exist```
Not sure why this happens, there should be no references to invisible rows anymore.

**Comment by bartcharbon** [comment]
If I show TypeTest as an admin, the dataexplorer shows 1 page and claims there are 20 rows, if I change the number of rows per page to 100 it shows 40 rows...

**Comment by tommydeboer** [comment]
**Issue:**
Give 'All Users' permission VIEW to Package 'sys'

```Message:Failed to convert from type [java.lang.String] to type [java.lang.Long] for value 'sys_idx'; nested exception is java.lang.NumberFormatException: For input string: "sys_idx"```

File: molgenis-data-platform/src/main/java/org/molgenis/data/meta/system/SystemEntityTypePersister.java [hunk 24]
```diff
@@ -56,6 +61,21 @@ public void persist()
...
+		{
+			String aclClass = EntityIdentityUtils.toType(systemEntityType);
+			// TODO remove casting
```
Comment (tommydeboer) [inline molgenis-data-platform/src/main/java/org/molgenis/data/meta/system/SystemEntityTypePersister.java]: resolve TODO

File: molgenis-data-security/src/main/java/org/molgenis/data/security/owned/RowLevelSecurityRepositoryDecorator.java [hunk 25]
```diff
@@ -0,0 +1,302 @@
...
+import java.util.stream.Stream;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.StreamSupport.stream;
+import static org.molgenis.data.security.EntityPermission.READ;
+import static org.molgenis.data.security.EntityPermission.WRITE;
+
+/**
+ * RepositoryDecorator that works on EntityType that extends OwnedEntityType.
```
Comment (tommydeboer) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/owned/RowLevelSecurityRepositoryDecorator.java]: Remove OwnedEntityType from docs

**Comment by tommydeboer** [comment]
Travis times out but I have confirmed that the tests succeed locally
